share/ci/pypicongpu_generator.py

```diff
@@ -0,0 +1,319 @@
+import requests
+import sys
+import yaml
+import re
+from typing import List, Dict, Callable
+import pkg_resources
+
+"""
+This file is part of the PIConGPU.
+Copyright 2023 PIConGPU contributors
+Authors: Simeon Ehrig
+License: GPLv3+
+"""
+
+"""@file Generate different CI test jobs for different Python version and
+         depending on a requirement.txt.
+
+Prints yaml code for a GitLab CI child pipeline to stout. The test parameter
+are split in two kinds of inputs. The Python versions to test are defined in
+the script, also the names of dependencies to test and it's test strategy. The
+version range of the dependencies to test are defined in a requirements.txt.
+The path of the requirements.txt is set via first application argument.
+
+First, the script reads the requirements.txt. If a dependency is marked to test
+in the script, it calculates the test versions. Therefore it downloads all
+available versions from pypi.org for each package. Afterwards it filters the
+versions via a filter strategy. For example, take all release versions or take
+each latest major version. Than the script removes all versions, which are not
+supported, defined in the requirements.txt. The result a complete list of all
+Python and dependencies versions to test.
+
+In the second part, the script creates the full combination matrix for all test
+versions and creates a CI job for each combination. Each job is printed on
+stdout.
+
+The number of combinations depends on:
+- number of supported Python version
+- number of dependencies to test
+- the test strategy of each dependencies to test
+- versions restrictions in the requirements.txt
+- releases of the dependencies
+
+@param First application argument: Path to the requirements.txt
+"""
+
+
+# caches the parsed dependencies of the requirements.txt here
+req_versions: Dict[str, pkg_resources.Requirement] = {}
+
+
+def cs(text: str, color: str) -> str:
+    """Print the text in a different color on the command line. The text after
+       the function has the default color of the command line.
+
+    Parameters
+    ----------
+        text (str): text to be colored
+        color (str): Name of the color. If wrong color or empty, use default
+                    color of the command line.
+
+    Returns
+    -------
+        str: text with bash pre and post string for coloring
+    """
+
+    if color is None:
+        return text
+
+    output = ""
+    if color == "Red":
+        output += "\033[0;31m"
+    elif color == "Green":
+        output += "\033[0;32m"
+    elif color == "Yellow":
+        output += "\033[1;33m"
+
+    return output + text + "\033[0m"
```

Comment:
this should probably become an imported library function in a separate file, code duplication is bad

Reply:
For normal software, I would agree. But that are standalone scripts. This scripts should be easily moveable, which is not possible if a dependency between two files exist. In our case we would already have the "complicated" situation, that `pypicongpu_generator.py` and `requirements_txt_modifier.py` shares the same code, but they are not located in the same folder.

To solve the problem correctly, we would need an installation script, which solves the problem. IMO, this is much more work, than keep the copied code persistent.

Reply:
@BrianMarre and I agreed in a personal discussion that my approach is fine for this use case.